src/components/common/PrivateRoute.js

```diff
@@ -0,0 +1,65 @@
+import React, { useState, useEffect } from 'react';
+import { connect } from 'react-redux';
+import { useHistory } from 'react-router-dom';
+import { SecureRoute, useOktaAuth } from '@okta/okta-react';
+import { authenticateUser } from '../../state/actions/auth/authenticateUser';
+import { getProfile } from '../../state/actions/userProfile/getProfile';
+
+const PrivateRoute = ({
+  component: Component,
+  loadingComponent: LoadingComponent = <></>, // TODO: create general page loading component
+  path,
+  redirect,
+  allowRoles, // should be an array of allowed role_id's i.e. [3, 4]
+  isAuthenticated,
+  profile_id,
+  userProfile,
+  dispatch,
+  ...rest
+}) => {
+  const { push } = useHistory();
+  const { authState, authService } = useOktaAuth();
+  const [loading, setLoading] = useState(true); // hiding contents
+
+  useEffect(() => {
+    if (Object.keys(userProfile).length === 0) {
+      if (profile_id === null) {
+        if (authState.isPending || authState.isAuthenticated) {
+          dispatch(authenticateUser(authState, authService));
+        } else {
+          push(redirect);
+        }
+      } else {
+        dispatch(getProfile(profile_id));
```

Comment:
Since getProfile is dispatched from within authenticateUser is it still needed here?

Reply:
awesome, the code is clean also I like how you are explain everything in the video.

Reply:
I moved the `getProfile` function into its own function in PR #[101](https://github.com/BloomTech-Labs/underdog-devs-fe-a/pull/101).